client/src/components/TopBar.jsx

```diff
@@ -0,0 +1,134 @@
+import React, { useState } from "react";
+import "./TopBar.css";
+
+// Validation utility function
+const isValidInput = (title, url) => {
+  const isValidTitle = typeof title === "string" && title.trim() !== "";
+  const isValidUrl =
+    typeof url === "string" &&
+    /^(https?:\/\/)?(www\.)?(youtube\.com\/watch\?v=|youtu\.be\/)[\w-]{11}($|&list=[\w-]+)/.test(
+      url
+    );
+  return isValidTitle && isValidUrl;
+};
+
+function TopBar({ onAddCard, cards, onOrderChange }) {
+  const [title, setTitle] = useState("");
+  const [url, setUrl] = useState("");
+  const [errorMessage, setErrorMessage] = useState("");
+  const [order, setOrder] = useState("desc");
+  const [fetching, setFetching] = useState(false);
+
+
+  const handleTitleChange = (event) => {
+    setTitle(event.target.value);
+  };
+
+  const handleUrlChange = (event) => {
+    setUrl(event.target.value);
+  };
+
+  const handleAddClick = () => {
+    if (!isValidInput(title, url)) {
+      setErrorMessage(
+        "Input URL or Title is invalid. Please provide a valid YouTube link."
+      );
+      setFetching(false);
+      return;
+    }
+
+    setFetching(true);
+
+    let videoCode = "";
+
+    if (url.includes("youtu.be/")) {
+      videoCode = url.split("youtu.be/")[1];
+    } else if (url.includes("&list=")) {
+      videoCode = url.split("watch?v=")[1].split("&")[0];
+    } else if (url.includes("watch?v=")) {
+      videoCode = url.split("watch?v=")[1];
+    } else {
+      setErrorMessage("Invalid YouTube URL format.");
+      return;
+    }
+
+    if (cards.length > 0) {
+      const isDuplicate = cards.some((card) => card.url === videoCode);
+      if (isDuplicate) {
+        setErrorMessage("This video is already added.");
+        setFetching(false);
+        return;
+      }
+    }
+
+    const newCard = {
+      title: title.trim(),
+      url: videoCode,
+    };
+
+    fetch("https://video-server-iiqf.onrender.com/videos", {
```

Comment:
When running the client and server locally, it'd be even better if you could feed some .env variables into the app to point the client to localhost if running locally, and to onrender if running in 'production' mode.

Reply:
can you please explain more about this?

Reply:
The line above has a hardcoded UR for the videos endpoint on onrender.com.
If you're running the code locally, it would be good to be able to point the app to the locally running server (on localhost) rather than the server on onrender.com.
Ideally you can do this without changing the code each time. 
Your app code could make a decision to point to either localhost or onrender.com using techniques like:
- change the app code to check the browser hostname - if it's localhost then make requests to the server on localhost. If the browser hostname is netlify, or wherever the app is deployed to, then make requests to the server at onrender.com
- use a .env file locally to load some environment variables into the app. You can read them with a library like dotenv. The .env file would be in your gitignore, so not in git, and would not be present when the app is running on Netlify. Then, your default could be to use the server at onrender.com, unless an environment variable is loaded that overrides that and points to the server on localhost

Reply:
I added env. thank you so much